DynamicShortcut makes app crash on startup

```
2019-04-24 15:08:38.997 com.squareup.cash.beta.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.squareup.cash.beta.debug, PID: 10131
    java.lang.RuntimeException: Unable to get provider leakcanary.internal.LeakSentryInstaller: java.lang.IllegalArgumentException: Max number of dynamic shortcuts exceeded
        at android.app.ActivityThread.installProvider(ActivityThread.java:6376)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:5932)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5847)
        at android.app.ActivityThread.access$1000(ActivityThread.java:198)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1637)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6649)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:826)
     Caused by: java.lang.IllegalArgumentException: Max number of dynamic shortcuts exceeded
        at android.os.Parcel.createException(Parcel.java:1946)
        at android.os.Parcel.readException(Parcel.java:1910)
        at android.os.Parcel.readException(Parcel.java:1860)
        at android.content.pm.IShortcutService$Stub$Proxy.addDynamicShortcuts(IShortcutService.java:497)
        at android.content.pm.ShortcutManager.addDynamicShortcuts(ShortcutManager.java:386)
        at leakcanary.internal.InternalLeakCanary.addDynamicShortcut(InternalLeakCanary.kt:141)
        at leakcanary.internal.InternalLeakCanary.onLeakSentryInstalled(InternalLeakCanary.kt:63)
        at leakcanary.internal.InternalLeakSentry.install(InternalLeakSentry.kt:61)
        at leakcanary.internal.LeakSentryInstaller.onCreate(LeakSentryInstaller.kt:19)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1917)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:1892)
        at android.app.ActivityThread.installProvider(ActivityThread.java:6371)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:5932) 
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5847) 
        at android.app.ActivityThread.access$1000(ActivityThread.java:198) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1637) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6649) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:826) 
     Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.pm.ShortcutService.enforceMaxActivityShortcuts(ShortcutService.java:1591)
        at com.android.server.pm.ShortcutPackage.enforceShortcutCountsBeforeOperation(ShortcutPackage.java:1104)
        at com.android.server.pm.ShortcutService.addDynamicShortcuts(ShortcutService.java:1876)
        at android.content.pm.IShortcutService$Stub.onTransact(IShortcutService.java:118)
        at android.os.Binder.execTransact(Binder.java:731)
```

--- a/leakcanary-android-core/src/main/java/leakcanary/internal/InternalLeakCanary.kt
+++ b/leakcanary-android-core/src/main/java/leakcanary/internal/InternalLeakCanary.kt
@@ -2,6 +2,7 @@ package leakcanary.internal
 
 import android.app.Application
 import android.app.Service
+import android.content.ComponentName
 import android.content.Context
 import android.content.Intent
 import android.content.pm.PackageInfo
@@ -19,6 +20,7 @@ import android.util.Log
 import com.squareup.leakcanary.core.BuildConfig
 import com.squareup.leakcanary.core.R
 import leakcanary.AnalysisResult
+import leakcanary.CanaryLog
 import leakcanary.GcTrigger
 import leakcanary.HeapDump
 import leakcanary.LeakCanary
@@ -74,10 +76,6 @@ internal object InternalLeakCanary : LeakSentryListener {
     val shortcutManager = application.getSystemService(ShortcutManager::class.java)!!
     val dynamicShortcuts = shortcutManager.dynamicShortcuts
 
-    if (dynamicShortcuts.size >= shortcutManager.maxShortcutCountPerActivity) {
-      return
-    }
-
     val shortcutInstalled =
       dynamicShortcuts.any { shortcut -> shortcut.id == DYNAMIC_SHORTCUT_ID }
 
@@ -89,6 +87,16 @@ internal object InternalLeakCanary : LeakSentryListener {
     mainIntent.addCategory(Intent.CATEGORY_LAUNCHER)
     mainIntent.setPackage(application.packageName)
     val activities = application.packageManager.queryIntentActivities(mainIntent, 0)
+        .filter {
+          it.activityInfo.name != "leakcanary.internal.activity.LeakLauncherActivity"
+        }
+
+    if (activities.isEmpty()) {
+      return
+    }
+
+    val firstMainActivity = activities.first()
+        .activityInfo
 
     // Displayed on long tap on app icon
     val longLabel: String
@@ -102,10 +110,8 @@ internal object InternalLeakCanary : LeakSentryListener {
       shortLabel = leakActivityLabel
     } else {
 
-      val activity = activities.first()
-
-      val firstLauncherActivityLabel = if (activity.activityInfo.labelRes != 0) {
-        application.getString(activity.activityInfo.labelRes)
+      val firstLauncherActivityLabel = if (firstMainActivity.labelRes != 0) {
+        application.getString(firstMainActivity.labelRes)
       } else {
         val applicationInfo = application.applicationInfo
         if (applicationInfo.labelRes != 0) {
@@ -130,15 +136,38 @@ internal object InternalLeakCanary : LeakSentryListener {
       }
     }
 
+    val componentName = ComponentName(firstMainActivity.packageName, firstMainActivity.name)
+
+    val shortcutCount = dynamicShortcuts.count { shortcutInfo ->
+      shortcutInfo.activity == componentName
+    } + shortcutManager.manifestShortcuts.count { shortcutInfo ->
+      shortcutInfo.activity == componentName
+    }
+
+    if (shortcutCount >= shortcutManager.maxShortcutCountPerActivity) {
+      return
+    }
+
     val intent = LeakActivity.createIntent(application)
     intent.action = "Dummy Action because Android is stupid"
     val shortcut = Builder(application, DYNAMIC_SHORTCUT_ID)
         .setLongLabel(longLabel)
         .setShortLabel(shortLabel)
+        .setActivity(componentName)
         .setIcon(Icon.createWithResource(application, R.mipmap.leak_canary_icon))
         .setIntent(intent)
         .build()
-    shortcutManager.addDynamicShortcuts(listOf(shortcut))
+
+    try {
+      shortcutManager.addDynamicShortcuts(listOf(shortcut))
+    } catch (ignored: Throwable) {
+      CanaryLog.d(
+          ignored,
+          "Could not add dynamic shortcut. " +
+              "shortcutCount=$shortcutCount, " +
+              "maxShortcutCountPerActivity=${shortcutManager.maxShortcutCountPerActivity}"
+      )
+    }
   }
 
   override fun onReferenceRetained() {

